include/ginkgo/core/base/lin_op.hpp

```diff
@@ -827,6 +827,66 @@ class EnableLinOp
     GKO_ENABLE_SELF(ConcreteLinOp);
 };
 
+/**
+ * The EnableValueTypeLinOp mixin adds several overloads to the EnableLinOp
+ * mixin. The added overloads allow using scalar arguments for the
+ * LinOp::apply() with four parameters. For more details see @ref EnableLinOp
+ */
+template <typename ConcreteLinOp, typename ValueType,
+          typename PolymorphicBase = LinOp>
+class EnableValueTypedLinOp
+    : public EnableLinOp<ConcreteLinOp, PolymorphicBase> {
+public:
+    using value_type = ValueType;
+
+    using EnableLinOp<ConcreteLinOp, PolymorphicBase>::EnableLinOp;
+    using EnableLinOp<ConcreteLinOp, PolymorphicBase>::apply;
+
+    const ConcreteLinOp *apply(const ValueType alpha, const LinOp *b,
+                               const ValueType beta, LinOp *x) const
+    {
+        auto dense_alpha = scalar_to_dense(alpha, this->get_executor());
+        auto dense_beta = scalar_to_dense(beta, this->get_executor());
+        return this->apply(dense_alpha.get(), b, dense_beta.get(), x);
+    }
```

Comment:
I think this part can be templated and put it in the linop directly?
and the scalar_to_dense is free function, so there is no cycle dep

Reply:
I've tried that, and it is possible. However, one problem I haven't figured out is how to handle `EnableLinOp`. If this is part of linop, it should be implemented again in enablelinop (or am I wrong with that?). This is more difficult, since the template needs to be instantiated explicitly somehow, and I don't know how to do this for a member function of an already templated class.

Reply:
I think you does not need to instantiate the template member function

Reply:
After some discussion with Tobias, I've decided to keep this in a separate mixin. Perhaps when linop gets some information about the used value type something like this could be implemented.

Reply:
Is there any reason for that?
or any downside from templated member function?
If user does not use the same type as LinOp, ginkgo will handle it via mix_and_match, I think?

Reply:
Only within a group (real or complex), but there is no implicit real -> complex or complex -> real conversion

Reply:
I think mix_and_match should be able to add the real->complex and limited usage for complex->real when the complex does not contain imag part.
does the mixin allow complex -> real?
I think the compiler does not allow std::complex<double>{3, 0} -> double?

Reply:
No, since all conversions need to be symmetric (conditionally inheriting from a type is really difficult), we can't represent this in the type hierarchy. Adding real-to-complex conversion to make_temporary_conversion would be possible though, I'll take a look at that. This would probably need to be implemented in `core/base/lin_op.cpp`, though I am not entirely happy with this function failing e.g. if you call `apply(1, x, 0, y)` with a linker error on the missing `apply(int, LinOp, int, LinOp)`.

Reply:
In current one, you also need to know which type of linop at least which valuetype of linop.
both cases you need to know about that.
it throws error when create dense with int.
if scalar_to_dense have a specialized to handle it, it should be okay.
integer will use float to represent.

Reply:
To make my example more concrete:
```cpp
    template <typename ValueType>
    void apply(const ValueType alpha, const LinOp *b,
                               const ValueType beta, LinOp *x) const
    {
        auto dense_alpha = scalar_to_dense(alpha, this->get_executor());
        auto dense_beta = scalar_to_dense(beta, this->get_executor());
        return this->apply(dense_alpha.get(), b, dense_beta.get(), x);
    }
    ...
    op->apply(1, x, 0, y);
```
causes a linker error because scalar_to_dense<int> is not instantiated.

On the other hand, in a non-templated ValueType overload
```cpp
   op->apply(1, x, 0, y);
```
casts 1 and 0 to ValueType and at worst gives a warning about the mismatching types.

I'm trying to avoid the use of unrestricted template member functions, since users don't get a compilation error at compile time, but only a much less useful error at link time complaining about a missing scalar_to_dense<int> symbol. I believe having the concrete type in here makes for much better type-safety in the common use case, which is using concrete types instead of generic LinOp types.

Reply:
the template one should be good for using.
```
auto dense_alpha = scalar_to_dense(int, this->get_executor);
// gives Dense<float> or Dense<double> with some addition on scalar_to_dense
```
it will not complain the scalar_to_dense<int>, when we add the conversion.
`op->apply(1, x, 0, y);` will gives implicitly conversion to ValueType, I think.

Reply:
the apply will accept different input when the object is concrete type or linop
what's this target for? for all user, the developer using ginkgo or ginkgo's developer.

Reply:
> it will not complain the scalar_to_dense, when we add the conversion.
> op->apply(1, x, 0, y); will gives implicitly conversion to ValueType, I think.

Yes, that is intended. Template type deduction requires the exact type to match, so scalar_to_dense(...) would try to use the non-existent `scalar_to_dense<int>` instantiation and cause a linker error. On the other hand, with an explicit non-template overload, even implicit conversions can be applied, so for a complex LinOp, you could have something like
`apply(1.0, x, std::complex<double>{0.0, 1.0}, y)` which works in the non-templated setting but doesn't compile in the templated setting.

> what's this target for? for all user, the developer using ginkgo or ginkgo's developer.

I would imagine that this is most useful for users and maybe example development. For generic library code and things like solvers, where the repeated allocations are not negligible, I would use explicit LinOps anyways.

> the apply will accept different input when the object is concrete type or linop

We have similar differences e.g. in the `apply` function, where `LinOp::apply` returns `LinOp*`, but `EnableLinOp::apply` returns a `ConcreteLinOp*`. More generally, only concrete types expose all functionality that the type has explicitly, which I would be totally fine with.

Reply:
maybe I put thing together such that it is misleading.
```
// for integer
Dense<float> scalar_to_dense<int>(...)
// for floating
Dense<float> scalar_to_dense<float>(...)
```
If we have this, the conversion should be okay.
current `op->apply(1, x, 0, y)` gives implicitly conversion to ValueType as templated one.
current pr conversion is on host but templated one may be on device.

`op->apply(1.0, x, std::complex<double>{0.0, 1.0}, y)` can be handled by two templated parameter for alpha and beta.

It can used when user only create the object and then use it.
If they do some combination, the type could be `LinOp*`

In my feeling, when I know concrete type, it is easy to use scalar_to_dense.
when I have concrete type, it is easy to know the valuetype and its executor, so I can easily to use these parameter to generate alpha/beta.
but when I have LinOp, I only have executor, so I need to use `dynamic_cast` through all possible types (and maybe more template params needs to know) or pick one valuetype to generate via `scalar_to_dense` thanks to `mixed_and_match`.

If we does not want `scalar_to_dense<int>` return `Dense<float>`, maybe the templated one needs to request the input always be floating point. Also, it will lead `op->apply(scalar_to_dense(1, ref), ...)` fail and `op->apply(1, ...)` work in current pr.
If we allow `scalar_to_dense<int>` return `Dense<float>`, it will be good for templated one.

It's good for example, but we start to have some LinOp like Multigrid without ValueType, so it will not always usable.
(Out of topic, I am thinking whether CG does really need ValueType because it just depends on the input matrix not the operation).
I think it is like the mixed_and_match place. It is a good tool to investigate ginkgo or develop something in ginkgo.

current pr:
pro: the conversion is host, do not contain any conversion on device.
con: needs to be concrete type or know the valuetype
templated:
pro: work for all linop
con: does not use the value type if it has one, so it might need conversion on device again.
If the linop does not implement the mixed_and_match, it will throw the link issue or runtime error?

Dense->apply return Dense and LinOp->apply return LinOP seems to be the same for me
I agree with you. only concreate type contains all functionalities but originally it's means different function name of the class to me. (maybe related to it, the naming issue `compute_absolute_linop` and `compute_absoulte`, sorry for out of topic again)